File: bfsdfs.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <list>
#include <queue>
#include <set>
#include <string>
#include <vector>

using namespace std;

typedef long long ll;
#define ln "\n"
#define mp make_pair
#define pb push_back
#define f first
#define s second
#define INF 2e18

void dfs(vector<vector<int>> &graph, vector<bool> &visited, int startNode) {
  visited[startNode] = true;
  cout << startNode << " ";
  for (int neighbour = 0; neighbour < graph.size(); neighbour++) {
    if (graph[startNode][neighbour] == 1 && !visited[neighbour]) {
      dfs(graph, visited, neighbour);
    }
  }
}

void bfs(vector<vector<int>> &graph, int startNode) {
  queue<int> q;
  vector<bool> visited(graph.size(), false);
  q.push(startNode);
  visited[startNode] = true;
  while (!q.empty()) {
    int u = q.front();
    q.pop();
    cout << u << " ";
    for (int neighbour = 0; neighbour < graph.size(); neighbour++) {
      if (graph[u][neighbour] == 1 && !visited[neighbour]) {
        q.push(neighbour);
        visited[neighbour] = true;
      }
    }
  }
}

int main() {
  // cout<<"Hello world !";
  int n;
  cin >> n;
  vector<vector<int>> graph(n, vector<int>(n, 0));
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      cin >> graph[i][j];
    }
  }
  vector<bool> visited(n, false);
  int startNode = 0;
  //   dfs(graph, visited, startNode);
  bfs(graph, startNode);
}

File: BfsDfs.java
import java.util.*;
import java.io.*;

class Solution {

    void dfs(int[][] graph, int node, ArrayList<Boolean> visited) {
        visited.set(node, true);
        System.out.print(node + " ");
        for (int neighbour = 0; neighbour < graph.length; neighbour++) {
            if (graph[node][neighbour] == 1 && visited.get(neighbour) != true) {
                dfs(graph, neighbour, visited);
            }
        }
        System.out.println();
    }

    void bfs(int[][] graph, int startNode, ArrayList<Boolean> visited) {
        Queue<Integer> q = new LinkedList<>();
        q.add(startNode);
        visited.set(startNode, true);
        while (!q.isEmpty()) {
            int node = q.poll();
            System.out.print(node + " ");
            for (int neighbour = 0; neighbour < graph.length; neighbour++) {
                if (graph[node][neighbour] == 1 && visited.get(neighbour) != true) {
                    q.add(neighbour);
                    visited.set(neighbour, true);
                }
            }
            System.out.println();

        }
    }
}

public class BfsDfs {
    public static void main(String[] args) {
        // System.out.println("Hello World!!!");
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        int[][] graph = new int[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                graph[i][j] = sc.nextInt();
            }
        }
        // for (int i = 0; i < n; i++) {
        // for (int j = 0; j < m; j++) {
        // System.out.print(graph[i][j] + " ");
        // }
        // System.out.println();
        // }
        Solution obj = new Solution();
        // int ans = obj.findCircleNum(graph);
        // System.out.println(ans);
        ArrayList<Boolean> visited = new ArrayList<Boolean>(n);
        for (int i = 0; i < n; i++) {
            visited.add(false);
        }
        // obj.dfs(graph, 0, visited);
        obj.bfs(graph, 0, visited);
    }
}